lib/charms/tempo_k8s/v2/tracing.py

```diff
@@ -902,9 +902,19 @@ def _get_endpoint(
     def get_endpoint(
         self, protocol: ReceiverProtocol, relation: Optional[Relation] = None
     ) -> Optional[str]:
-        """Receiver endpoint for the given protocol."""
+        """Receiver endpoint for the given protocol.
+
+        Raises:
+        ProtocolNotRequestedError:
+            If the charm unit is the leader unit and attempts to obtain an endpoint for a protocol it did not request.
+        """
         endpoint = self._get_endpoint(relation or self._relation, protocol=protocol)
         if not endpoint:
+            # non-leader units cannot access local app databag
+            if not self._charm.unit.is_leader():
```

Comment:
It's always a of a trade-off, but I wonder if in this case it wouldn't be better to raise an exception if we are not the leader and force the caller to guard instead.
Reason is that the caller should be aware that only leaders can see the endpoints, and now if this call returns None, we don't know if it's because there is no endpoint (yet) or because we're not allowed to see it.

Reply:
hmmm good point
but, will it make a difference to know the reason why we got None? If there's an issue with the protocol sent, you'll still get notified, as the leader will crash. 

In the case of the non-leader, having None will atleast not break just because of a race condition. Another reason for me is that its not quite easy to catch the issue in terms of how often it'd happen. 

Its 50 50 for me :D

Reply:
Well presumably the user will unittest this, and if you don't set leader, you'll get an error as soon as you call get_endpoint, and you'll know you need a guard. Which by the way, is what happens already, only now you would get a better error.
So I think it's more in line with the current behaviour to keep raising if you're calling this without leadership

Reply:
I'll keep the existing behavior, but will include more info about this in the docstring of the function